src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed partial class RedirectHandler : HttpMessageHandler
+    {
+        private readonly HttpMessageHandler _initialInnerHandler;       // Used for initial request
+        private readonly HttpMessageHandler _redirectInnerHandler;      // Used for redirects; this allows disabling auth
+        private readonly int _maxAutomaticRedirections;
+
+        public RedirectHandler(int maxAutomaticRedirections, HttpMessageHandler initialInnerHandler, HttpMessageHandler redirectInnerHandler)
+        {
+            Debug.Assert(initialInnerHandler != null);
+            Debug.Assert(redirectInnerHandler != null);
+
+            if (maxAutomaticRedirections < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxAutomaticRedirections));
+            }
+
+            _maxAutomaticRedirections = maxAutomaticRedirections;
+            _initialInnerHandler = initialInnerHandler;
+            _redirectInnerHandler = redirectInnerHandler;
+        }
+
+        protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            HttpResponseMessage response = await _initialInnerHandler.SendAsync(request, cancellationToken).ConfigureAwait(false);
+
+            uint redirectCount = 0;
+            Uri redirectUri;
+            while ((redirectUri = GetUriForRedirect(request.RequestUri, response)) != null)
+            {
+                response.Dispose();
+
+                redirectCount++;
+                if (redirectCount > _maxAutomaticRedirections)
+                {
+                    throw new HttpRequestException(SR.net_http_max_redirects);
+                }
+
+                // Clear the authorization header.
+                request.Headers.Authorization = null;
+
+                // Set up for the redirect
+                request.RequestUri = redirectUri;
+                if (RequestRequiresForceGet(response.StatusCode, request.Method))
+                {
+                    request.Method = HttpMethod.Get;
+                    request.Content = null;
```

Comment:
What happens if request.Headers.TransferEncodingChunked was set to true?  Will we now end up throwing on the redirect because we've nulled out the content?

Reply:
Good question.  I'll add a test and fix if necessary.

Reply:
What happens is that we incorrectly send Transfer-Encoding: chunked on the redirected request.

I've pushed a fix and the new tests.

Looks like CurlHandler is failing this as well, so I'll file an issue and disable tests.

Reply:
Filed #27301